splitter/splitter.go

```diff
@@ -0,0 +1,218 @@
+	"sync"
+	"time"
+
+	events "github.com/bluesky-social/indigo/events"
+	"github.com/bluesky-social/indigo/events/schedulers/sequential"
+	"github.com/bluesky-social/indigo/models"
+	"github.com/gorilla/websocket"
+	logging "github.com/ipfs/go-log"
+)
+
+var log = logging.Logger("splitter")
+
+type Splitter struct {
+	Host   string
+	erb    *EventRingBuffer
+	events *events.EventManager
+}
+
+func NewSplitter(host string, persister events.EventPersistence) *Splitter {
+	erb := NewEventRingBuffer(20000, 1000)
+
+	em := events.NewEventManager(erb)
+	return &Splitter{
+		Host:   host,
+		erb:    erb,
+		events: em,
+	}
+}
+
+func (s *Splitter) Start() error {
+	return nil
+}
+
+func sleepForBackoff(b int) time.Duration {
+	if b == 0 {
+		return 0
+	}
+
+	if b < 50 {
+		return time.Millisecond * time.Duration(rand.Intn(100)+(5*b))
+	}
+
+	return time.Second * 5
+}
+
+func (s *Splitter) subscribeWithRedialer(ctx context.Context, host string, cursor int64) {
+	d := websocket.Dialer{}
+
+	protocol := "wss"
+
+	var backoff int
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		url := fmt.Sprintf("%s://%s/xrpc/com.atproto.sync.subscribeRepos?cursor=%d", protocol, host, cursor)
+		con, res, err := d.DialContext(ctx, url, nil)
+		if err != nil {
+			log.Warnw("dialing failed", "host", host, "err", err, "backoff", backoff)
+			time.Sleep(sleepForBackoff(backoff))
+			backoff++
+
+			continue
+		}
+
+		log.Info("event subscription response code: ", res.StatusCode)
+
+		if err := s.handleConnection(ctx, host, con, &cursor); err != nil {
+			log.Warnf("connection to %q failed: %s", host, err)
+		}
+	}
+}
+
+func (s *Splitter) handleConnection(ctx context.Context, host string, con *websocket.Conn, lastCursor *int64) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	sched := sequential.NewScheduler("splitter", s.events.AddEvent)
+	return events.HandleRepoStream(ctx, con, sched)
+}
+
+func sequenceForEvent(evt *events.XRPCStreamEvent) int64 {
+	switch {
+	case evt.RepoCommit != nil:
+		return evt.RepoCommit.Seq
+	case evt.RepoHandle != nil:
+		return evt.RepoHandle.Seq
+	case evt.RepoMigrate != nil:
+		return evt.RepoMigrate.Seq
+	case evt.RepoTombstone != nil:
+		return evt.RepoTombstone.Seq
+	case evt.RepoInfo != nil:
+		return -1
+	default:
+		return -1
+	}
+}
+
+func NewEventRingBuffer(chunkSize, nchunks int) *EventRingBuffer {
+	return &EventRingBuffer{
+		chunkSize:     chunkSize,
+		maxChunkCount: nchunks,
+	}
+}
+
+type EventRingBuffer struct {
+	lk            sync.Mutex
+	chunks        []*ringChunk
+	chunkSize     int
+	maxChunkCount int
+
+	broadcast func(*events.XRPCStreamEvent)
+}
+
+type ringChunk struct {
+	lk  sync.Mutex
+	buf []*events.XRPCStreamEvent
+}
+
+func (rc *ringChunk) append(evt *events.XRPCStreamEvent) {
+	rc.lk.Lock()
+	defer rc.lk.Unlock()
+	rc.buf = append(rc.buf, evt)
+}
+
+func (rc *ringChunk) events() []*events.XRPCStreamEvent {
+	rc.lk.Lock()
+	defer rc.lk.Unlock()
+	return rc.buf
```

Comment:
Does this need to be a copy of the buffer?

Reply:
nope, the buffer is append-only, and this operation takes a copy of the *slice*, which means we can read from it in a threadsafe manner all of the data that exists in the slice at this point in time. Future appends will not impact this slice we've grabbed (even if they end up triggering a resize of the slice, the old memory is still valid and safe to read from)